File: csharp/src/Algorithms/BFS.cs
namespace DSA.Algorithms;

/// <summary>
/// Breadth-First Search implementation.
/// Time: O(V + E)
/// Space: O(V)
/// </summary>
public static class BFS
{
    public static List<T> Traverse<T>(Graph<T> graph, T start, Action<T>? callback = null) where T : notnull
    {
        var result = new List<T>();

        if (!graph.HasVertex(start))
            return result;

        var visited = new HashSet<T>();
        var queue = new Queue<T>();

        queue.Enqueue(start);
        visited.Add(start);

        while (queue.Count > 0)
        {
            var current = queue.Dequeue();
            result.Add(current);
            callback?.Invoke(current);

            foreach (var neighbor in graph.Neighbors(current))
            {
                if (!visited.Contains(neighbor))
                {
                    visited.Add(neighbor);
                    queue.Enqueue(neighbor);
                }
            }
        }

        return result;
    }

    public static List<T>? FindPath<T>(Graph<T> graph, T start, T end) where T : notnull
    {
        if (!graph.HasVertex(start) || !graph.HasVertex(end))
            return null;

        if (EqualityComparer<T>.Default.Equals(start, end))
            return new List<T> { start };

        var visited = new HashSet<T>();
        var parent = new Dictionary<T, T>();
        var queue = new Queue<T>();

        queue.Enqueue(start);
        visited.Add(start);

        while (queue.Count > 0)
        {
            var current = queue.Dequeue();

            if (EqualityComparer<T>.Default.Equals(current, end))
            {
                var path = new List<T>();
                var node = end;
                while (!EqualityComparer<T>.Default.Equals(node, start))
                {
                    path.Add(node);
                    node = parent[node];
                }
                path.Add(start);
                path.Reverse();
                return path;
            }

            foreach (var neighbor in graph.Neighbors(current))
            {
                if (!visited.Contains(neighbor))
                {
                    visited.Add(neighbor);
                    parent[neighbor] = current;
                    queue.Enqueue(neighbor);
                }
            }
        }

        return null;
    }

    public static Dictionary<T, int> Distances<T>(Graph<T> graph, T start) where T : notnull
    {
        var distances = new Dictionary<T, int>();

        if (!graph.HasVertex(start))
            return distances;

        var visited = new HashSet<T>();
        var queue = new Queue<(T vertex, int distance)>();

        queue.Enqueue((start, 0));
        visited.Add(start);
        distances[start] = 0;

        while (queue.Count > 0)
        {
            var (current, dist) = queue.Dequeue();

            foreach (var neighbor in graph.Neighbors(current))
            {
                if (!visited.Contains(neighbor))
                {
                    visited.Add(neighbor);
                    distances[neighbor] = dist + 1;
                    queue.Enqueue((neighbor, dist + 1));
                }
            }
        }

        return distances;
    }
}

File: typescript/src/algorithms/bfs.ts
/**
 * Breadth-First Search implementation.
 * Time: O(V + E)
 * Space: O(V)
 */

import { Graph } from './graph';

export function bfs<T>(graph: Graph<T>, start: T, callback?: (vertex: T) => void): T[] {
  const result: T[] = [];

  if (!graph.hasVertex(start)) {
    return result;
  }

  const visited = new Set<T>();
  const queue: T[] = [start];
  visited.add(start);

  while (queue.length > 0) {
    const current = queue.shift()!;
    result.push(current);

    if (callback) {
      callback(current);
    }

    for (const neighbor of graph.neighbors(current)) {
      if (!visited.has(neighbor)) {
        visited.add(neighbor);
        queue.push(neighbor);
      }
    }
  }

  return result;
}

export function bfsFindPath<T>(graph: Graph<T>, start: T, end: T): T[] | undefined {
  if (!graph.hasVertex(start) || !graph.hasVertex(end)) {
    return undefined;
  }

  if (start === end) {
    return [start];
  }

  const visited = new Set<T>();
  const parent = new Map<T, T>();
  const queue: T[] = [start];
  visited.add(start);

  while (queue.length > 0) {
    const current = queue.shift()!;

    if (current === end) {
      const path: T[] = [];
      let node: T | undefined = end;
      while (node !== undefined && node !== start) {
        path.push(node);
        node = parent.get(node);
      }
      path.push(start);
      return path.reverse();
    }

    for (const neighbor of graph.neighbors(current)) {
      if (!visited.has(neighbor)) {
        visited.add(neighbor);
        parent.set(neighbor, current);
        queue.push(neighbor);
      }
    }
  }

  return undefined;
}

export function bfsDistances<T>(graph: Graph<T>, start: T): Map<T, number> {
  const distances = new Map<T, number>();

  if (!graph.hasVertex(start)) {
    return distances;
  }

  const visited = new Set<T>();
  const queue: Array<{ vertex: T; distance: number }> = [{ vertex: start, distance: 0 }];
  visited.add(start);
  distances.set(start, 0);

  while (queue.length > 0) {
    const { vertex: current, distance: dist } = queue.shift()!;

    for (const neighbor of graph.neighbors(current)) {
      if (!visited.has(neighbor)) {
        visited.add(neighbor);
        distances.set(neighbor, dist + 1);
        queue.push({ vertex: neighbor, distance: dist + 1 });
      }
    }
  }

  return distances;
}
